### ISSUE TITLE
MiMo kernel correctness gate: test-backend-ops cases at exact MiMo shapes + logit-drift check

### ISSUE THREAD

**Issue by dymoo** [dymoo/mimo-halo-lab#101]
Part of #93 · Kernel family: #40, #41 · Pins: `manifests/upstreams.json` (strix-llama.cpp `8c1c282e`)

## Goal

Build the correctness gate that every MiMo kernel change must pass before anyone makes a performance claim. Today `test-backend-ops` has no cases at MiMo's real shapes. It also can't compare the provisional winner's expert count, 141 (REAP45), against the fallback's 154 (REAP40). So "numerical equivalence proven" (#40 acceptance) has nothing concrete to run against. This ticket defines the cases and the end-to-end drift check once, and every kernel ticket reuses them.

## Inputs

- MiMo dims from the pinned manifest (`manifests/models/mimo-v2.6-flash-rl`): hidden 4096, `moe_intermediate_size` 2048, top-8 sigmoid+bias router, Q/K head dim 192, V head dim 128, 64 query heads, 8 KV heads on SWA layers (window 128) and 4 on global layers (GQA 8 / 16), 48 layers (47 MoE).
- Candidate formats from `configs/experiments/compression-sweep.json`: expert payloads are native MXFP4 (`mxfp4_native`, 4.25 bpw). Stage-one triage (`docs/experiments/compression-sweep.md`) ranks REAP45 (141 experts) provisional quality-first and REAP40 (154) as the fallback.
- The pinned tree's `tests/test-backend-ops.cpp` (`-o`, `-p`, `perf` mode) and `llama-perplexity --kl-divergence-base / --kl-divergence`.

## Outputs

- A tracked patch `patches/strix-mimo-backend-ops-cases.patch`, with base and sha256 recorded in `manifests/upstreams.json` like the other strix patches. It adds MiMo-shape cases:
  - `MUL_MAT_ID` MXFP4: n_expert ∈ {141, 154, 256}, n_used 8, gate/up 4096→2048 and down 2048→4096, n_tokens ∈ {1, 2, 4, 8, 16, 64, 512, 2048}.
  - Fused top-k MoE routing: sigmoid gating + `exp_probs_b` bias, n_expert ∈ {141, 154, 256}, k = 8.
  - `FLASH_ATTN_EXT` Dk=192 / Dv=128, GQA 8 and 16, KV length ∈ {128, 4096, 32768}, KV type ∈ {f16, q8_0}, with sinks as the model uses them.
- A drift-check procedure in `docs/hardware-bringup.md`, cited by later tickets. It saves reference logits from the unoptimised pinned build with `--kl-divergence-base`, then scores a candidate build with `--kl-divergence` on a fixed, non-golden text slice. It records mean/p99 KLD and top-1 agreement.
- A recorded noise floor: KLD between the two unoptimised lanes (Vulkan vs HIP with `HIP_LAUNCH_BLOCKING=1`). Kernel changes are gated at **≤ 2× this floor**, and the multiplier is written down alongside the floor.

## Blockers

None for case definition and the Mac/CPU run. The gfx1151 run of these cases happens inside the profiling ticket (the kernel-profile issue that depends on this one).

## Acceptance

- [ ] Patch applies cleanly at the strix pin, is hashed in `manifests/upstreams.json`, and `test-backend-ops -o MUL_MAT_ID` / `-o FLASH_ATTN_EXT` / the topk-moe op list the new cases.
- [ ] New cases pass against the CPU reference on a non-Halo build (Mac), with logs/checksums committed as evidence (sanitized, no local paths).
- [ ] The drift-check procedure is documented with exact commands and a fixed text slice that is not golden and not calibration (hash recorded).
- [ ] The noise-floor measurement is marked as pending hardware, not invented.
- [ ] No Qwen-specific paths touched.

## Commands

```bash
# after applying the patch + building (see docs/hardware-bringup.md §C):
./build/bin/test-backend-ops -o MUL_MAT_ID -p 'n_mats=(141|154|256)'
./build/bin/test-backend-ops -o FLASH_ATTN_EXT -p 'hsk=192'
# drift check (hardware day):
./build/bin/llama-perplexity -m <gguf> -f <fixed-slice> --kl-divergence-base <ref.kld>
./build/bin/llama-perplexity -m <gguf> -f <fixed-slice> --kl-divergence-base <ref.kld> --kl-divergence
```

(The `-p` regexes are illustrative. Match whatever `vars()` strings the new cases print.)

## Provenance

Patch + sha256 in `manifests/upstreams.json`, and test logs under `manifests/runtime/`. No weights, prompts or raw logits in Git.
